services/rfq/relayer/inventory/rebalance.go

```diff
@@ -27,183 +27,185 @@ type RebalanceManager interface {
 //
 //nolint:cyclop,nilnil
-func getRebalanceAmount(span trace.Span, cfg relconfig.Config, tokens map[int]map[common.Address]*TokenMetadata, originTokenData, destTokenData *TokenMetadata) (amount *big.Int, err error) {
+func getRebalanceAmount(ctx context.Context, cfg relconfig.Config, tokens map[int]map[common.Address]*TokenMetadata, rebalance *RebalanceData) (amount *big.Int, err error) {
+	span := trace.SpanFromContext(ctx)
+
 	// get the maintenance and initial values for the destination chain
-	maintenancePct, err := cfg.GetMaintenanceBalancePct(destTokenData.ChainID, destTokenData.Addr.Hex())
+	maintenancePct, err := cfg.GetMaintenanceBalancePct(rebalance.DestMetadata.ChainID, rebalance.DestMetadata.Addr.Hex())
 	if err != nil {
 		return nil, fmt.Errorf("could not get maintenance pct: %w", err)
 	}
-	initialPct, err := cfg.GetInitialBalancePct(destTokenData.ChainID, destTokenData.Addr.Hex())
+	initialPct, err := cfg.GetInitialBalancePct(rebalance.DestMetadata.ChainID, rebalance.DestMetadata.Addr.Hex())
 	if err != nil {
 		return nil, fmt.Errorf("could not get initial pct: %w", err)
 	}
 
 	// calculate maintenance threshold relative to total balance
-	tokenName := originTokenData.Name
-	totalBalance := big.NewInt(0)
-	for _, tokenMap := range tokens {
-		for _, tokenData := range tokenMap {
-			if tokenData.Name == tokenName {
-				totalBalance.Add(totalBalance, tokenData.Balance)
-			}
-		}
-	}
+	totalBalance := getTotalBalance(cfg, tokens, rebalance.OriginMetadata.Name, rebalance.Method)
 	maintenanceThresh, _ := new(big.Float).Mul(new(big.Float).SetInt(totalBalance), big.NewFloat(maintenancePct/100)).Int(nil)
 	if span != nil {
 		span.SetAttributes(attribute.Float64("maintenance_pct", maintenancePct))
 		span.SetAttributes(attribute.Float64("initial_pct", initialPct))
-		span.SetAttributes(attribute.String("max_token_balance", originTokenData.Balance.String()))
-		span.SetAttributes(attribute.String("min_token_balance", destTokenData.Balance.String()))
+		span.SetAttributes(attribute.String("max_token_balance", rebalance.OriginMetadata.Balance.String()))
+		span.SetAttributes(attribute.String("min_token_balance", rebalance.DestMetadata.Balance.String()))
 		span.SetAttributes(attribute.String("total_balance", totalBalance.String()))
 		span.SetAttributes(attribute.String("maintenance_thresh", maintenanceThresh.String()))
 	}
 
 	// no need to rebalance if we are not below maintenance threshold on destination
-	if destTokenData.Balance.Cmp(maintenanceThresh) > 0 {
+	if rebalance.DestMetadata.Balance.Cmp(maintenanceThresh) > 0 {
+		fmt.Println("returning nil")
 		return nil, nil
 	}
 
 	// calculate the amount to rebalance vs the initial threshold on destination
 	initialThresh, _ := new(big.Float).Mul(new(big.Float).SetInt(totalBalance), big.NewFloat(initialPct/100)).Int(nil)
-	amount = new(big.Int).Sub(originTokenData.Balance, initialThresh)
+	amount = new(big.Int).Sub(rebalance.OriginMetadata.Balance, initialThresh)
```

Comment:
@dwasse could you explain the logic here wrt initial value of the `amount`? Is the happy path here to rebalance exactly as much to get the destination chain balance back to the initial?

Reply:
I reworked this function slightly and added some documentation:

```
// At a high level, there are three steps:
//  1. Evaluate the relevant maintenance / initial thresholds on origin and destination
//  2. Arrive at an initial rebalance amount, first targeting the initial threshold on origin,
//     and then clipping by the initial threshold on destination if necessary
//  3. Filter the rebalance amount by the configured min and max
```

In short, happy path is to bridge whatever amount takes the origin to initial threshold. If this number would take destination above initial threshold, we crop by that initial threshold as well